### PR TITLE
Change the state machine used for processing new blocks

### LINKED ISSUE #13: [Critical] TokenLocker allows to use header that did not pass challenge period
As reported by @abacabadabacaba 

> The state of the bridge is queried using blockHashes and blockMerkleRoots functions. However, these functions return info about a new block during the challenge period, during which the block should not be trusted yet.

This allows malicious actor to submit an invalid header and while it is challenged perform invalid operations.

Interestingly, it was present in https://github.com/near/rainbow-bridge-cli/tree/2fca92c314405b344b48359165dd262d2b910b72 and security audit did not catch it.

### PR DESCRIPTION
**Pull request by abacabadabacaba**
Now a newly added block that is not yet trusted is stored in untrustedHead. After the challenge period is over, it can be moved to head. Fixes #13, #15.

### CODE REVIEW

File: nearbridge/contracts/NearBridge.sol
Comment (ailisp) [inline nearbridge/contracts/NearBridge.sol]: Great to make a difference between head and untrustedHead!

**Review by ailisp — COMMENTED** [review]
Looks good, need bump version in package.json if we want to use this immediately. verify_near_headers test failed because it simply try to submit all blocks during a period, so refused by "not sufficiently new" check, either change test or remove some blocks for now

**Comment by abacabadabacaba** [comment]
I fixed verify_near_headers, but it still won't pass cli tests because there is new constructor parameter.

**Review by maxzaver — CHANGES_REQUESTED** [review]
This logic makes contract more complex by adding an additional public method. Instead we can make `head` and `backupHead` private and add an additional method `getTrustedHead()` which returns either `head` or `backupHead` depending on whether `validAfter` has passed.

**Comment by abacabadabacaba** [comment]
> This logic makes contract more complex by adding an additional public method. Instead we can make head and `backupHead` private and add an additional method `getTrustedHead()` which returns either `head` or `backupHead` depending on whether `validAfter` has passed.

I think that the logic is now less complex. There is `head` which is trusted. A new method is needed to move `untrustedHead` to `head` when the challenge period is over without pushing a new block. This also makes this block available to `blockHashes` and `blockMerkleRoots`. It is possible to add an extra method, but it won't make things less complex, because there should be still a way to commit a block so that `NearProver` can use it (it uses `blockMerkleRoots`).

Or I can just make `commitBlock` internal, so a block will only become available when another block is submitted on top of it.

Overall, this code needs a lot of refactoring, but for now I try not to change it too much.

**Comment by maxzaver** [comment]
> This also makes this block available to blockHashes and blockMerkleRoots. It is possible to add an extra method, but it won't make things less complex, because there should be still a way to commit a block so that NearProver can use it (it uses blockMerkleRoots).

We could add `trustedMerkleRoots` method that filter the last merkle root if the header has not passed the challenge period yet. By adding `commitBlock` we are delegating complexity to the contract user who now need to call an extra method before being able to use this contract.

> Or I can just make commitBlock internal, so a block will only become available when another block is submitted on top of it.

This will make header that passed challenge period not useful until the next header is submitted. Ideally, we want header to be useful as soon as it passes the challenge period.

**Comment by abacabadabacaba** [comment]
> We could add `trustedMerkleRoots` method that filter the last merkle root if the header has not passed the challenge period yet.

That's not a good API design. "Trusted" should be the default (and indeed the only) option.

Anyway, I changed the API to return the last block as soon as the challenge period is over.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol]: It looks like we should remove this check since it should never be `none`, because `_checkBp` wouldn't allow such block to be submitted.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol]: FYI this changes the definition of the `hash` field to an actual header hash. I assume it is intended.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol]: This change requires changes to https://github.com/near/rainbow-bridge-lib and https://github.com/near/rainbow-bridge-cli (we need to add a new CLI arg). Please create corresponding PRs, without them the bridge will be broken.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:69]: Please add a comment saying something like: "when it is zero `untrustedHead`, `untrustedHeadIsFromNextEpoch`, `untrustedNextBlockProducers` should be considered to be not set".

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:304]: Please add a comment saying that if block is from the next epoch it is required to have next block producer so `nextBlockProducers` will never be overridden with empty collection.

**Review by maxzaver — APPROVED** [review]
Looks good. There are some suggestions to add comments. We also need to add PR for rainbow-bridge-cli and rainbow-bridge-lib to make the tests pass. Sorry for the delay.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol]: Yes, this value was never actually used, so I repurposed this field.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol]: I really wish the code to deploy the contracts were in the same repository as the contracts themselves. Currently, I cannot merge this PR because it fails CI.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol]: Why? We only need signatures from 2/3 of the validators, so some could be missing.

**Comment by abacabadabacaba** [comment]
@nearmax I addressed the comments and also fixed another bug, is everything OK now? Will make a PR for near-bridge-cli and -lib a bit later. I can't merge this PR because of CI.

**Comment by abacabadabacaba** [comment]
Created pull requests: near/rainbow-bridge-cli#348, near/rainbow-bridge-lib#14.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [reply nearbridge/contracts/NearBridge.sol]: You are right, I did not read the code carefully.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [reply nearbridge/contracts/NearBridge.sol]: I agree. CC @Kouprin

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:237]: Was this the additional bug that you fixed? It is hard to see because of the forced push.

**Review by maxzaver — APPROVED** [review]
Approving, assuming https://github.com/near/rainbow-bridge-sol/pull/20/files?file-filters%5B%5D=.sol#diff-b55dd2aebd3038c032230f5c9710455dR212 is the only non-comment change since previous review.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol:237]: The bug that I fixed was that I added a check to `checkBlockProducerSignatureInHead` that `signatureIndex` is valid.

**Review by ailisp — APPROVED** [review]
Looks good!

File: dist.sh
Comment (maxzaver) [inline dist.sh:19]: Could we add a comment explaining what is this for?

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol]: This PR might be a good opportunity to make sure we follow Solidity style guide: https://solidity.readthedocs.io/en/v0.6.0/style-guide.html#avoiding-naming-collisions

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:104]: AFAIU `override` is only useful when contract inherits from another contract, not from an interface.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:199]: It is unclear why do we not return `numBlockProducers` when there is an untrusted head.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol]: Could we please add comments explaining each field?

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [reply nearbridge/contracts/NearBridge.sol:237]: Are you talking about `signatureIndex < untrustedBlockProducers.bpsLength` check? Because it seems `!untrustedApprovals[signatureIndex].none` check does not fix vulnerabilities.

File: nearbridge/contracts/NearBridge.sol
Comment (maxzaver) [inline nearbridge/contracts/NearBridge.sol:240]: The numbers in the comments were useful because they corresponded to the spec: https://nomicon.io/ChainSpec/LightClient.html

**Review by maxzaver — CHANGES_REQUESTED** [review]

File: dist.sh
Comment (abacabadabacaba) [reply dist.sh:19]: This was because of nomiclabs/truffle-flattener#55. Will add a comment.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol:104]: In Solidity 0.6, `override` is _mandatory_ when inheriting from an interface. Without it, it just won't compile.

File: nearbridge/contracts/NearBridge.sol
Comment (abacabadabacaba) [reply nearbridge/contracts/NearBridge.sol:199]: We do return the number of block producers when there is an untrusted head. This number is used by the watchdog to determine the number of signatures to check.

**Review by maxzaver — APPROVED** [review]
Looks good

**Comment by maxzaver** [comment]
I have tested this change locally, so I am going to merge it and publish the package.

**Pull request merged**
